### ISSUE TITLE
Improve testing API to support the continuous integration scenario

### ISSUE THREAD

**Issue by Sanjo** [meteor/meteor#6809]
From my experience there are two scenarios how you run tests:

1. In development
  * Tests rerun automatically after you make changes to the app
  * Developer gets rich feedback on the test results

2. On continuous integration server
  * Tests run only once
  * Results are outputted on stdout
  * Test runner (meteor test) exits after test run:
    *  with 0 when all tests passed
    *  with non-zero when tests failed

I think the current testing API is mainly designed for the first scenario.
Please improve the testing API to also support the second scenario.

__Proposed additions__

1. The user can run tests in the specified mode.
    * I think a --ci flag would do it. So you could use `meteor test --ci` to run app tests in CI mode.
2. Testing driver packages can handle both cases.
    * I think Meteor should tell the driver package in which mode the tests should run. Meteor could execute runTests with: `runTests({ci: true})` __on the server side__ in CI mode. As alternative a new separate export could be introduced. E.g. `runTestsInCi`.
    * The runTests function can return a boolean. true = all tests passed, false = some tests failed.
      * In CI mode Meteor will exit
        * with status code 0 when runTests returned true
        * with status code 1 when runTests return false.

**Comment by Sanjo** [comment]
Related to https://github.com/meteor/meteor/issues/6755.
And relevant for @tmeasday.

**Comment by tmeasday** [comment]
I think we can do something pretty simply for this. @Sanjo would it make sense to close this issue and discuss on #6755?

I would be happy to review PRs on this once we decide on the design.

**Comment by Sanjo** [comment]
> would it make sense to close this issue and discuss on #6755?

I don't think so, because in this issue I ask for a differentiation between running in CI and while developing locally.

In issue #6755 Aldeed asks for a differentiation between running tests once or continuous. So I ask for stronger semantics, so I can do the things listed in my initial post based on the scenario the tests run in.

Also it is important that runTests is executed on the server side in CI mode, because there is no user who will open the browser with the tests for you.

**Comment by tmeasday** [comment]
I guess I mean let's use a single ticket to discuss this related suite of changes to the test running infrastructure to support `--once` (and/or `--ci`) better.

**Comment by Sanjo** [comment]
Yeah, if the --once flag that @aldeed suggested is also for the purpose of running the tests in CI, we can discuss this in one ticket.

One concern right now for me is if it is really necessary to create different driver-packages for different reporters and CI/non-ci. Or rather have one driver package for a testing framework, that can handle different configurations.

Also I would prefer a unified way for driver package for how to do things like running on CI. This is why I would like to reflect it in the Meteor testing API.

**Comment by aldeed** [comment]
@Sanjo Currently, with 1.3, there is `runTests` on the server already, but it's called `start`. So the way the `dispatch:` test driver packages currently break down is as follows:
- `dispatch:mocha`: Runs ONLY server tests, by exporting a `start` function on the server. Can be used on CI by adding `--once` or locally by adding `TEST_WATCH=1`. 
- `dispatch:mocha-phantomjs`: Runs server AND client tests, by exporting a `start` function on the server. The `start` function kicks off client tests through phantomjs and shows those in the server console. Can be used on CI by adding `--once` or locally by adding `TEST_WATCH=1`.
- `dispatch:mocha-browser`: Runs server tests by exporting a `start` function on the server. Runs client tests by exporting a `runTests` function on the client. You load in browser to run client tests and see the results. Can't be used on CI.

#6755 proposal would remove the need for `TEST_WATCH=1`, allowing `--once` to put it in CI mode. I don't know that a `--ci` alias is necessary, but it could be done.

So in theory I think you're suggesting that a single pkg that combines `mocha-phantomjs` with `mocha-browser` would be better (or the jasmine version of that in your case). I don't have a strong opinion, but I think I lean slightly toward keeping everything separate, mainly because it keeps dependencies trim. Already you could use `dispatch:mocha-phantomjs` alone and it works on CI as well as locally in watch mode. The only catch is that you see the results in the server console and they run in phantom, which I know in some cases is less desirable than running them in Chrome. But it should be possible to make a similar pkg that runs them in Chrome with Selenium.

**Comment by Sanjo** [comment]
@aldeed Your proposal sounds good to me. Let's get --once in.

I'm closing this in favor of https://github.com/meteor/meteor/issues/6755.
